src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -493,10 +497,32 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
 
     private function createTypeNotFoundMessage($type, $label)
     {
-        if (!$classOrInterface = class_exists($type, false) ? 'class' : (interface_exists($type, false) ? 'interface' : null)) {
+        $autowireById = Definition::AUTOWIRE_BY_ID === $this->currentDefinition->getAutowired();
+        if (!$classOrInterface = class_exists($type, $autowireById) ? 'class' : (interface_exists($type, false) ? 'interface' : null)) {
             return sprintf('Cannot autowire %s for service "%s": Class or interface "%s" does not exist.', $label, $this->currentId, $type);
         }
-        $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        if (null === $this->types) {
+            $this->populateAvailableTypes();
+        }
+        if (!$autowireById) {
+            $message = sprintf('No services were found matching the "%s" %s and it cannot be auto-registered', $type, $classOrInterface);
+        } elseif (isset($this->ambiguousServiceTypes[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to any of these existing services: "%s"', $type, implode('", "', $this->ambiguousServiceTypes[$type]));
+        } elseif (isset($this->types[$type])) {
+            $message = sprintf('Service "%s" does not exist. This type could be aliased to the existing "%s" service', $type, $this->types[$type]);
```

Comment:
It's perhaps outside the scope of this PR, but we could also add this "alias" recommendation to the "other" mode, when there are multiple definitions (https://github.com/symfony/symfony/blob/daac6c24490cfd9eb995ce6fb69095b7241f8b13/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php#L56)